security/advancedtls/advancedtls.go

```diff
@@ -0,0 +1,377 @@
+	Certificates []tls.Certificate
+	// If GetClientCertificate is set and Certificates is nil, the server will invoke this
+	// function every time asked to present certificates to the client when a new connection is
+	// established. This is known as peer certificate reloading.
+	GetCertificate func(*tls.ClientHelloInfo) (*tls.Certificate, error)
+	RootCertificateOptions
+	// If the server want the client to send certificates.
+	RequireClientCert bool
+}
+
+func (o *ClientOptions) config() (*tls.Config, error) {
+	if o.RootCACerts == nil && o.GetRootCAs == nil && o.VerifyPeer == nil {
+		return nil, fmt.Errorf(
+			"client needs to provide root CA certs, or a custom verification function")
+	}
+	// We have to set InsecureSkipVerify to true to skip the default checks and use the
+	// verification function we built from buildVerifyFunc.
+	config := &tls.Config{
+		ServerName:           o.ServerNameOverride,
+		Certificates:         o.Certificates,
+		GetClientCertificate: o.GetClientCertificate,
+		RootCAs:              o.RootCACerts,
+		InsecureSkipVerify:   true,
+	}
+	return config, nil
+}
+
+func (o *ServerOptions) config() (*tls.Config, error) {
+	if o.Certificates == nil && o.GetCertificate == nil {
+		return nil, fmt.Errorf("either Certificates or GetCertificate must be specified")
+	}
+	if o.RequireClientCert && o.GetRootCAs == nil && o.RootCACerts == nil {
+		return nil, fmt.Errorf("server needs to provide root CA certs if requiring client cert")
+	}
+	clientAuth := tls.NoClientCert
+	if o.RequireClientCert {
+		// We fall back to normal config settings if users don't need to reload root certificates.
+		// If using RequireAndVerifyClientCert, the underlying stack would use the default
+		// checking and ignore the verification function we built from buildVerifyFunc.
+		// If using RequireAnyClientCert, the code would skip all the checks and use the
+		// function from buildVerifyFunc.
+		if o.RootCACerts != nil {
+			clientAuth = tls.RequireAndVerifyClientCert
+		} else {
+			clientAuth = tls.RequireAnyClientCert
+		}
+	}
+	config := &tls.Config{
+		ClientAuth:     clientAuth,
+		Certificates:   o.Certificates,
+		GetCertificate: o.GetCertificate,
+	}
+	if o.RootCACerts != nil {
+		config.ClientCAs = o.RootCACerts
+	}
+	return config, nil
+}
+
+// advancedTLSCreds is the credentials required for authenticating a connection using TLS.
+type advancedTLSCreds struct {
+	config     *tls.Config
+	verifyFunc CustomVerificationFunc
+	getRootCAs func(rawConn net.Conn, rawCerts [][]byte) (*x509.CertPool, error)
+	isClient   bool
+}
+
+func (c advancedTLSCreds) Info() credentials.ProtocolInfo {
+	return credentials.ProtocolInfo{
+		SecurityProtocol: "tls",
+		SecurityVersion:  "1.2",
+		ServerName:       c.config.ServerName,
+	}
+}
+
+func (c *advancedTLSCreds) ClientHandshake(ctx context.Context, authority string,
+	rawConn net.Conn) (_ net.Conn, _ credentials.AuthInfo, err error) {
```

Comment:
Nit: I wouldn't name return values at all unless you need to or they help with the documentation.

Also, I'd unwrap the line as long lines are fine in Go and this isn't that excessive.